### DISCUSSION TITLE
Shell, Navigation and Page instantiation

### DISCUSSION THREAD

**Discussion by ioiooi** [dotnet/maui#5810] [Q&A]
Hello,

are Pages that are defined in the visual hierarchy only ever instantiated _once_? And if that is the case is there a way to clear the cache and force the creation of a new Page instance the next time I navigate to it?
  
For example the following hierachy is described in a `AppShell.xaml`: 

```
<ShellContent Route="LoginPage" Shell.FlyoutBehavior="Disabled" ContentTemplate="{DataTemplate local:LoginPage}" />

<FlyoutItem FlyoutDisplayOptions="AsMultipleItems">
    <ShellContent Title="About" Icon="tab_about.png" Route="AboutPage" ContentTemplate="{DataTemplate local:AboutPage}" />
    <ShellContent Title="Browse" Icon="tab_feed.png" ContentTemplate="{DataTemplate local:ItemsPage}" />
</FlyoutItem>

<ShellContent Route="SecretPage" Shell.FlyoutBehavior="Disabled" ContentTemplate="{DataTemplate local:SecretPage}" />
```
  
The LoginPage constructor gets called when the App is started. Navigate to the AboutPage with `Shell.Current.GoToAsync($"//{nameof(AboutPage)}")` and its constructor gets called.

Now lets say the AboutPage has a Logout-Button which does some things and at the end it calls `Shell.Current.GoToAsync($"//{nameof(LoginPage)}")`. The LoginPage constructor does not get called. Navigate back from the LoginPage to the AboutPage and its constructor also does not get called.

The same deal if I would navigate from the LoginPage to the SecretPage. The constructor of SecretPage would only get called the first time.

Lets say I use ViewModels for all those Pages. I can't really create a state in the ViewModel constructor because the Page only gets instantiated once and so in my case the ViewModel also only gets instantiated once. I would need to do the work in a OnAppearing.

I would want for the Pages after the LoginPage to always be newly constructed in order to make sure that theres a clean state.
  
PS: 
A lot of examples also contain DetailsPages... something like ItemsDetails, MonkeyDetails, CatDetails etc... Those Pages are never declared in the `AppShell.xaml` but are registered like so 
```
Routing.RegisterRoute("monkeys/details", typeof(MonkeyDetailPage));
```
and from what Ive seen seem to be always instantiated every single time theyre navigated to.

**Comment by NickDevoctomy** [comment]
Hello, I was just wondering if you had managed to find a solution for this? I could also do with handing page instantiation this way.  Have you tried asking in the MAUI Q&A site?

https://docs.microsoft.com/en-us/answers/topics/460228/dotnet-maui.html

**Reply by ioiooi** [reply]
Hi, nope sorry, I haven't found a solution for this.
  
In my case I just want to be sure that everything starts fresh after a Logout and that there are no residual objects or whatever from the previous User/Session. My workaround for this specific case is to create a new AppShell after a Logout. It looks something like this:
  
```
public void Logout()
{
    Application.Current!.MainPage = new AppShell();
}
```
  
cheers

**Reply by ederbond** [reply]
Thanks @ioiooi your workaround seems to work fine for me, I'm just not sure if it could let us to any other issue like a memory.

**Reply by Hooterr** [reply]
Has anyone looked into this, if it causes any memory issues?

**Reply by MV10** [reply]
That Q&A site is awful...

@Hooterr see [issue 9300](https://github.com/dotnet/maui/issues/9300); somebody from MS confirms that shouldn't leak.

There is also a workaround in [this](https://github.com/dotnet/maui/issues/9300#issuecomment-1416893792) comment.

**Comment by shadowfoxish** [comment]
Heh, I had the same question in a more roundabout way
https://github.com/dotnet/maui/discussions/16295
